### PR TITLE
Add documentation of data logs and the data log tool

### PR DESCRIPTION
**Pull request by PeterJohnson**

### CODE REVIEW

File: source/docs/software/telemetry/telemetry.rst [hunk 1]
```diff
@@ -25,3 +27,225 @@ WPILib supports several :ref:`dashboards <docs/software/dashboards/index:Dashboa
 Telemetry data can be sent to a WPILib dashboard using an associated WPILib method (for more details, see the documentation for the individual dashboard in question), or by :ref:`directly publishing to NetworkTables <docs/software/networktables/networktables-intro:Writing a simple NetworkTables program>`.
 
 While NetworkTables does not yet support serialization of complex data types (this is tentatively scheduled for 2023), *mutable* types from user code can be easily extended to interface directly with WPILib dashboards via the ``Sendable`` interface, whose usage is described in the next article.
+
+On-Robot Telemetry Recording Into Data Logs
```
[the remark this answers is not in the document]
Comment (PeterJohnson) [reply source/docs/software/telemetry/telemetry.rst]: Done.

**Review by Daltz333 — COMMENTED** [review]
I love this and I love the documentation here. One small note, all the references to button labels should be formatted like `:guilabel:'mylabel'` where you replace the apostrophe with `

File: source/docs/software/telemetry/datalog.rst [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+------------------------------------------
+
+The ``DataLogManager`` class (`Java <https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/wpilibj/DataLogManager.html>`__, `C++ <https://first.wpi.edu/wpilib/allwpilib/docs/release/cpp/classfrc_1_1_data_log_manager.html>`__) provides a centralized data log that provides automatic data log file management.  It automatically cleans up old files when disk space is low and renames the file based either on current date/time or (if available) competition match number.  The deta file will be saved to a USB flash drive if one is attached, or to ``/home/lvuser`` otherwise.
+
+Log files are initially named ``FRC_TBD_{random}.wpilog`` until the DS connects.  After the DS connects, the log file is renamed to ``FRC_yyyyMMdd_HHmmss.wpilog`` (where the date/time is UTC).  If the FMS is connected and provides a match number, the log file is renamed to ``FRC_yyyyMMdd_HHmmss_{event}_{match}.wpilog``.
+
+On startup, all existing ``FRC_TBD`` log files are deleted.  If there is less than 50 MB of free space on the target storage, ``FRC_`` log files are deleted (oldest to newest) until there is 50 MB free OR there are 10 files remaining.
```
Comment (Daltz333) [inline source/docs/software/telemetry/datalog.rst:22]: This is slightly confusing? Perhaps rewording this sentence to 
```suggestion
On startup, all existing log files where a DS has not been connected will be deleted.  If there is less than 50 MB of free space on the target storage, ``FRC_`` log files are deleted (oldest to newest) until there is 50 MB free OR there are 10 files remaining.
```

File: source/docs/software/telemetry/datalog.rst [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+        // (alternatively) Record only DS control data
+        DriverStation::StartDataLog(DataLogManager::GetLog(), false);
+
+Custom Data Logging using DataLog
+---------------------------------
+
+The ``DataLog`` class (`Java <https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/util/datalog/DataLog.html>`__, `C++ <https://first.wpi.edu/wpilib/allwpilib/docs/release/cpp/classwpi_1_1log_1_1_data_log.html>`__) and its associated LogEntry classes (e.g. ``BooleanLogEntry``, ``DoubleLogEntry``, etc) provides low-level access for writing data logs.
+
+.. note: Unlike NetworkTables, there is no change checking performed.  **Every** call to a LogEntry.append() function will result in a record being written to the data log.  Checking for changes and only appending to the log when necessary is the responsibility of the caller.
```
Comment (Daltz333) [inline source/docs/software/telemetry/datalog.rst:88]:
```suggestion
.. note: Unlike NetworkTables, there is no change checking performed.  **Every** call to a ``LogEntry.append()`` function will result in a record being written to the data log.  Checking for changes and only appending to the log when necessary is the responsibility of the caller.
```

**Comment by PeterJohnson** [comment]
Addressed comments.  The links to first.wpi.edu docs still need to be checked; the 2022.4.1 docs haven't published yet so it's uncertain if they're correct.

File: source/docs/software/telemetry/datalog.rst [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+Data logs consist of a series of timestamped records.  Control records allow starting, finishing, or changing the metadata of entries, and data records record data value changes.
+
+.. note: For more information on the details of the data log file format, see the `WPILib Data Log File Format Specification <https://github.com/wpilibsuite/allwpilib/blob/main/wpiutil/doc/datalog.adoc>`__.
+
+Standard Data Logging using DataLogManager
+------------------------------------------
+
+The ``DataLogManager`` class (`Java <https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/wpilibj/DataLogManager.html>`__, `C++ <https://first.wpi.edu/wpilib/allwpilib/docs/release/cpp/classfrc_1_1_data_log_manager.html>`__) provides a centralized data log that provides automatic data log file management.  It automatically cleans up old files when disk space is low and renames the file based either on current date/time or (if available) competition match number.  The deta file will be saved to a USB flash drive if one is attached, or to ``/home/lvuser`` otherwise.
```
Comment (sciencewhiz) [inline source/docs/software/telemetry/datalog.rst:18]: Should we say a fat32 formatted flash drive?

Comment (Daltz333) [reply source/docs/software/telemetry/datalog.rst:18]: Does exfat not also work?

Comment (sciencewhiz) [reply source/docs/software/telemetry/datalog.rst:18]: I was primarily meaning to exclude NTFS. I haven't tried exfat

Comment (sciencewhiz) [reply source/docs/software/telemetry/datalog.rst:18]: It appears that exfat doesn't get mounted

Comment (PeterJohnson) [reply source/docs/software/telemetry/datalog.rst:18]: Added a note box regarding USB flash drive formatting requirements.

**Review by sciencewhiz — APPROVED** [review]

**Review by Daltz333 — APPROVED** [review]

**Pull request merged**